File: android/src/main/java/com/ReactNativeBlobUtil/ReactNativeBlobUtilConfig.java
package com.ReactNativeBlobUtil;

import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

class ReactNativeBlobUtilConfig {

    public Boolean fileCache;
    public Boolean transformFile;
    public String path;
    public String appendExt;
    public ReadableMap addAndroidDownloads;
    public Boolean trusty;
    public Boolean wifiOnly = false;
    public String targetHostIp;
    public String key;
    public String mime;
    public Boolean auto;
    public Boolean overwrite = true;
    public long timeout = 60000;
    public Boolean increment = false;
    public Boolean followRedirect = true;
    public ReadableArray binaryContentTypes = null;
    public List<String> customCACerts = null;
    public List<String> pinnedHosts = null;
    public Boolean trustSystemCerts = false;

    ReactNativeBlobUtilConfig(ReadableMap options) {
        if (options == null)
            return;
        this.fileCache = options.hasKey("fileCache") && options.getBoolean("fileCache");
        this.transformFile = options.hasKey("transformFile") ? options.getBoolean("transformFile") : false;
        this.path = options.hasKey("path") ? options.getString("path") : null;
        this.appendExt = options.hasKey("appendExt") ? options.getString("appendExt") : "";
        this.trusty = options.hasKey("trusty") && options.getBoolean("trusty");
        this.wifiOnly = options.hasKey("wifiOnly") && options.getBoolean("wifiOnly");
        this.targetHostIp = options.hasKey("targetHostIp") ? options.getString("targetHostIp") : "";
        if (options.hasKey("addAndroidDownloads")) {
            this.addAndroidDownloads = options.getMap("addAndroidDownloads");
        }
        if (options.hasKey("binaryContentTypes"))
            this.binaryContentTypes = options.getArray("binaryContentTypes");
        if (this.path != null && path.toLowerCase(Locale.ROOT).contains("?append=true")) {
            this.overwrite = false;
        }
        if (options.hasKey("overwrite"))
            this.overwrite = options.getBoolean("overwrite");
        if (options.hasKey("followRedirect")) {
            this.followRedirect = options.getBoolean("followRedirect");
        }
        this.key = options.hasKey("key") ? options.getString("key") : null;
        this.mime = options.hasKey("contentType") ? options.getString("contentType") : null;
        this.increment = options.hasKey("increment") && options.getBoolean("increment");
        this.auto = options.hasKey("auto") && options.getBoolean("auto");
        if (options.hasKey("timeout")) {
            this.timeout = options.getInt("timeout");
        }
        if (options.hasKey("customCACerts")) {
            ReadableArray certsArray = options.getArray("customCACerts");
            if (certsArray != null && certsArray.size() > 0) {
                this.customCACerts = new ArrayList<>();
                for (int i = 0; i < certsArray.size(); i++) {
                    this.customCACerts.add(certsArray.getString(i));
                }
            }
        }
        if (options.hasKey("pinnedHosts")) {
            ReadableArray hostsArray = options.getArray("pinnedHosts");
            if (hostsArray != null && hostsArray.size() > 0) {
                this.pinnedHosts = new ArrayList<>();
                for (int i = 0; i < hostsArray.size(); i++) {
                    this.pinnedHosts.add(hostsArray.getString(i));
                }
            }
        }
        if (options.hasKey("trustSystemCerts")) {
            this.trustSystemCerts = options.getBoolean("trustSystemCerts");
        }
    }

}

File: android/src/main/java/com/ReactNativeBlobUtil/ReactNativeBlobUtilConfig.kt
package com.ReactNativeBlobUtil

import com.facebook.react.bridge.ReadableArray
import com.facebook.react.bridge.ReadableMap
import java.util.Locale

/**
 * The fetch config passed from JS, read once into fields.
 *
 * The fields keep the boxed, nullable types of the Java class: with null options
 * the constructor returns early and leaves most of them null, and the request code
 * reads them as it always has.
 */
internal class ReactNativeBlobUtilConfig(options: ReadableMap?) {

    var fileCache: Boolean? = null
    var transformFile: Boolean? = null
    var path: String? = null
    var appendExt: String? = null
    var addAndroidDownloads: ReadableMap? = null
    var trusty: Boolean? = null
    var wifiOnly: Boolean? = false
    var targetHostIp: String? = null
    var key: String? = null
    var mime: String? = null
    var auto: Boolean? = null
    var overwrite: Boolean? = true
    var timeout: Long = 60000
    var increment: Boolean? = false
    var followRedirect: Boolean? = true
    var binaryContentTypes: ReadableArray? = null
    var customCACerts: MutableList<String?>? = null
    var pinnedHosts: MutableList<String?>? = null
    var trustSystemCerts: Boolean? = false

    init {
        if (options != null) {
            fileCache = options.hasKey("fileCache") && options.getBoolean("fileCache")
            transformFile = if (options.hasKey("transformFile")) options.getBoolean("transformFile") else false
            path = if (options.hasKey("path")) options.getString("path") else null
            appendExt = if (options.hasKey("appendExt")) options.getString("appendExt") else ""
            trusty = options.hasKey("trusty") && options.getBoolean("trusty")
            wifiOnly = options.hasKey("wifiOnly") && options.getBoolean("wifiOnly")
            targetHostIp = if (options.hasKey("targetHostIp")) options.getString("targetHostIp") else ""
            if (options.hasKey("addAndroidDownloads")) {
                addAndroidDownloads = options.getMap("addAndroidDownloads")
            }
            if (options.hasKey("binaryContentTypes")) {
                binaryContentTypes = options.getArray("binaryContentTypes")
            }
            val currentPath = path
            if (currentPath != null && currentPath.lowercase(Locale.ROOT).contains("?append=true")) {
                overwrite = false
            }
            if (options.hasKey("overwrite")) {
                overwrite = options.getBoolean("overwrite")
            }
            if (options.hasKey("followRedirect")) {
                followRedirect = options.getBoolean("followRedirect")
            }
            key = if (options.hasKey("key")) options.getString("key") else null
            mime = if (options.hasKey("contentType")) options.getString("contentType") else null
            increment = options.hasKey("increment") && options.getBoolean("increment")
            auto = options.hasKey("auto") && options.getBoolean("auto")
            if (options.hasKey("timeout")) {
                timeout = options.getInt("timeout").toLong()
            }
            if (options.hasKey("customCACerts")) {
                val certsArray = options.getArray("customCACerts")
                if (certsArray != null && certsArray.size() > 0) {
                    customCACerts = MutableList(certsArray.size()) { certsArray.getString(it) }
                }
            }
            if (options.hasKey("pinnedHosts")) {
                val hostsArray = options.getArray("pinnedHosts")
                if (hostsArray != null && hostsArray.size() > 0) {
                    pinnedHosts = MutableList(hostsArray.size()) { hostsArray.getString(it) }
                }
            }
            if (options.hasKey("trustSystemCerts")) {
                trustSystemCerts = options.getBoolean("trustSystemCerts")
            }
        }
    }
}
